pkg/meter/native/vec.go

```diff
@@ -0,0 +1,146 @@
+// Licensed to Apache Software Foundation (ASF) under one or more contributor
+// license agreements. See the NOTICE file distributed with
+// this work for additional information regarding copyright
+// ownership. Apache Software Foundation (ASF) licenses this file to you under
+// the Apache License, Version 2.0 (the "License"); you may
+// not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing,
+// software distributed under the License is distributed on an
+// "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+// KIND, either express or implied.  See the License for the
+// specific language governing permissions and limitations
+// under the License.
+
+// Package native provides a simple meter system for metrics. The metrics are aggregated by the meter provider.
+package native
+
+import (
+	"context"
+	"strings"
+	"sync"
+	"time"
+
+	"github.com/robfig/cron/v3"
+	"google.golang.org/protobuf/types/known/timestamppb"
+
+	"github.com/apache/skywalking-banyandb/api/data"
+	commonv1 "github.com/apache/skywalking-banyandb/api/proto/banyandb/common/v1"
+	measurev1 "github.com/apache/skywalking-banyandb/api/proto/banyandb/measure/v1"
+	modelv1 "github.com/apache/skywalking-banyandb/api/proto/banyandb/model/v1"
+	"github.com/apache/skywalking-banyandb/banyand/queue"
+	"github.com/apache/skywalking-banyandb/pkg/bus"
+	"github.com/apache/skywalking-banyandb/pkg/convert"
+	"github.com/apache/skywalking-banyandb/pkg/logger"
+	"github.com/apache/skywalking-banyandb/pkg/meter"
+	pbv1 "github.com/apache/skywalking-banyandb/pkg/pb/v1"
+	"github.com/apache/skywalking-banyandb/pkg/timestamp"
+)
+
+const (
+	writeTimeout = 5 * time.Second
+)
+
+type metricWithLabelValues struct {
+	labelValues []string
+	metric      float64
+}
+
+type metricVec struct {
+	scheduler   *timestamp.Scheduler
+	pipeline    queue.Client
+	scope       meter.Scope
+	metrics     map[uint64]metricWithLabelValues
+	measureName string
+	mutex       sync.Mutex
+}
+
+func newMetricVec(measureName string, pipeline queue.Client, scope meter.Scope) *metricVec {
+	clock, _ := timestamp.GetClock(context.TODO())
+	n := &metricVec{
+		measureName: measureName,
+		pipeline:    pipeline,
+		scope:       scope,
+		scheduler:   timestamp.NewScheduler(log, clock),
+		metrics:     map[uint64]metricWithLabelValues{},
+	}
+	err := n.scheduler.Register("flush metrics", cron.Descriptor, "@every 5s", func(_ time.Time, _ *logger.Logger) bool {
```

Comment:
Please move this to "service.go". I prefer to keep a single scheduler for flushing all metrics.